### ISSUE TITLE
Use `if let` instead of `if matches!`

### ISSUE THREAD

**Issue by GideonBear** [rust-lang/rust-clippy#16660]
### What it does

Warn against use of `if matches!` where it could be written as `if let`. I don't know in which cases this is or isn't safe.

This should probably be a pedantic lint (see drawbacks).

### Advantage

- More readable
- Shorter
- More idiomatic

### Drawbacks

- The pattern and the value are switched around, which might've been done intentionally for readability.

### Example

```rust
enum E {
    A,
    B,
}

fn main() {
    let x = E::A;
    if matches!(x, E::A) {
        println!("It's an A");
    }
}
```

Could be written as:

```rust
enum E {
    A,
    B,
}

fn main() {
    let x = E::A;
    if let E::A = x {
        println!("It's an A");
    }
}
```
And:
```rust
enum E {
    A(u32),
    B,
}

fn main() {
    let x = E::A(1);
    if matches!(x, E::A(n) if n == 1) {
        println!("It's an A(1)");
    }
}
```
Could be written as:
```rust
enum E {
    A(u32),
    B,
}

fn main() {
    let x = E::A(1);
    if let E::A(n) = x && n == 1 {
        println!("It's an A(1)");
    }
}
```
(As long as let chains are available in the current Rust version)

**Comment by samueltardieu** [comment]
This would probably be a `restriction` lint, as the goal is to restrict the use of `if matches!(…)`. And some people prefer not to use `if let` if there no binding being created and used after the `if let` expression, so the reverse lint could also be created.

**Comment by sagudev** [comment]
We need to be sure there is no variable use in the block:
```rust
let n = 5;
if matches!(x, E::A(n) if n == 1) {
        println!("It's an A(1)");
        dbg!(n); // prints 5
}
```
but
```rust
let n = 5;
if let E::A(n) = x && n == 1 {
        println!("It's an A(1)");
        dbg!(n); // prints 1
}
```

**Comment by LeoniePhiline** [comment]
`if let` suggests a pattern extraction. This leads to confusion in cases like the `if let E::A = x` example above, where nothing is extracted, but also in the `if let E::A(n) = x && n == 1` case, where `n` is extracted but not used in the block.

I would argue that such use is unidiomatic:

`let` is a declaration. Pattern-matches which do not extract a binding used in the following block should always use `if matches!`. It reads naturally and clearly indicates that the extraction (if any) is not used beyond the scrutinee/condition expression.

To my mind, clippy should suggest `if matches!(x, E::A)` for `if let E::A = x`, and `if matches!(x, E::A(n) if n == 1)` for `if let E::A(n) = x && n == 1`, at least in the `pedantic` group.

This would be exactly the opposite of the lint as proposed.

**Comment by GideonBear** [comment]
IMO both versions should be in the restriction group, unless there is a clear consensus on which is better, which I think is impossible primarily because both can each be clearer in different situations (e.g. because the pattern and value are switched around).

**Comment by tmke8** [comment]
Another vote from me for preferring

```rust
if let E::A = x {
    println!("It's an A");
}
```

over

```rust
if matches!(x, E::A) {
    println!("It's an A");
}
```

Mainly because

- It looks more like native syntax as opposed to a function-like call.
- With if-let-chains I think it's natural to interpret `let PAT = EXPR` as a kind of boolean statement and so why shouldn't it be used for non-binding patterns like `E::A` as well? The [RFC on if-let-chains](https://rust-lang.github.io/rfcs/2497-if-let-chains.html#the-main-alternatives) also considered the syntax `EXPR is PAT`, which in this case would be `x is E::A`, which would clearly be superior to `matches!(x, E::A)`. In the end it was decided (correctly, I think) not to use `EXPR is PAT` for if-let-chains, but I would argue that we should still use `if let E::A = x` because it's spiritually supposed to be the same as `x is E::A`. In other words: we shouldn't let the decision against `EXPR is PAT` get in the way of using `if let` for non-binding patterns.
